### ISSUE TITLE
mc sometimes hangs on directory change with subshell on macOS

### ISSUE THREAD

**Issue by mc-butler** [MidnightCommander/mc#4071]
> [!IMPORTANT]
> This issue was migrated from Trac:
>
> | **Origin** | https://midnight-commander.org/ticket/4071 |
> | :------------- | :------------------------------------------------------------------------------------------------ |
> | **Reporter** | `olfway` (olfway@….com) |

<p>
I use mc 4.8.24 on mac os 10.15.3<br />
</p>

```
❯ env LC_MESSAGES=C /opt/mc/bin/mc -V
GNU Midnight Commander unknown
Built with GLib 2.64.1
Using the S-Lang library with terminfo database
With builtin Editor
With subshell support as default
With support for background operations
With mouse support on xterm
With internationalization support
With multiple codepages support
Virtual File Systems: cpiofs, tarfs, sfs, extfs, ftpfs, fish
Data types: char: 8; int: 32; long: 64; void *: 64; size_t: 64; off_t: 64;
```

```
❯ /opt/mc//bin/mc --configure-options
 '--prefix' '/opt/mc' '--without-x' '--with-screen=slang' '--disable-doxygen-html' '--disable-doxygen-dot' '--disable-doxygen-doc' 'CFLAGS=-O0 -g -ggdb' 'LDFLAGS=-L/usr/local/opt/gettext/lib -L/usr/local/opt/gettext/lib' 'CPPFLAGS=-I/usr/local/opt/gettext/include -I/usr/local/opt/gettext/include'
```
<p>
fish shell, version 3.1.0<br />
</p>
<p>
Sometimes, then I press enter to change directory mc just hangs.<br />
Also, there is a zombie kill process after that<br />
</p>
<p>
Backtrace from mc:<br />
</p>
<p>
 <br />
</p>

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff73a3a3b2 libsystem_kernel.dylib`__sigsuspend + 10
    frame #1: 0x00000001002f9394 mc`synchronize at common.c:497:9
    frame #2: 0x00000001002f81ba mc`feed_subshell(how=0, fail_on_error=0) at common.c:609:13
    frame #3: 0x00000001002f87e8 mc`do_subshell_chdir(vpath=0x00007fa164c0f8b0, update_prompt=0) at common.c:1345:5
    frame #4: 0x0000000100276299 mc`subshell_chdir(vpath=0x00007fa164c0f8b0) at panel.c:3234:9
    frame #5: 0x00000001002745f6 mc`_do_panel_cd(panel=0x00007fa164c0ec70, new_dir_vpath=0x00007fa164c0f570, cd_type=cd_exact) at panel.c:3275:5
    frame #6: 0x00000001002744e3 mc`do_panel_cd(panel=0x00007fa164c0ec70, new_dir_vpath=0x00007fa164c0f570, cd_type=cd_exact) at panel.c:4628:9
    frame #7: 0x00000001002758be mc`do_cd(new_dir_vpath=0x00007fa164c0f570, exact=cd_exact) at panel.c:5028:11
    frame #8: 0x0000000100279e76 mc`do_enter_on_file_entry(fe=0x00000001008670b8) at panel.c:2795:14
    frame #9: 0x000000010027865e mc`do_enter(panel=0x00007fa164c0ec70) at panel.c:2855:12
    frame #10: 0x00000001002765c1 mc`panel_execute_cmd(panel=0x00007fa164c0ec70, command=1) at panel.c:3446:9
    frame #11: 0x00000001002763f4 mc`panel_key(panel=0x00007fa164c0ec70, key=10) at panel.c:3608:20
    frame #12: 0x0000000100272655 mc`panel_callback(w=0x00007fa164c0ec70, sender=0x0000000000000000, msg=MSG_KEY, parm=10, data=0x0000000000000000) at panel.c:3688:16
    frame #13: 0x000000010023075a mc`send_message(w=0x00007fa164c0ec70, sender=0x0000000000000000, msg=MSG_KEY, parm=10, data=0x0000000000000000) at widget-common.h:216:15
    frame #14: 0x0000000100231a16 mc`dlg_key_event(h=0x00007fa164f05d90, d_key=10) at dialog.c:489:19
    frame #15: 0x0000000100231439 mc`dlg_process_event(h=0x00007fa164f05d90, key=10, event=0x00007ffeef9f5640) at dialog.c:1134:9
    frame #16: 0x0000000100231d48 mc`frontend_dlg_run(h=0x00007fa164f05d90) at dialog.c:545:9
    frame #17: 0x0000000100231b8e mc`dlg_run(h=0x00007fa164f05d90) at dialog.c:1167:5
    frame #18: 0x000000010026d8ed mc`do_nc at midnight.c:1836:16
    frame #19: 0x000000010020bb92 mc`main(argc=1, argv=0x00007ffeef9f5808) at main.c:405:21
    frame #20: 0x00007fff738d97fd libdyld.dylib`start + 1
    frame #21: 0x00007fff738d97fd libdyld.dylib`start + 1
(lldb) frame variable
(lldb) up
frame #1: 0x00000001002f9394 mc`synchronize at common.c:497:9
   494
   495 	    /* Wait until the subshell has stopped */
   496 	    while (subshell_alive && !subshell_stopped)
-> 497 	        sigsuspend (&old_mask);
   498
   499 	    if (subshell_state != ACTIVE)
   500 	    {
(lldb) frame variable
(sigset_t) sigchld_mask = 524288
(sigset_t) old_mask = 0

```
<p>
Backtrace from fish (part of):<br />
</p>

```
frame #2: 0x00000001012380fe fish`exec_external_command(parser=0x00007fb2e2d02030, j=std::__1::shared_ptr<job_t>::element_type @ 0x00007fb2e2f0f3c8 strong=2 weak=1, p=0x00007fb2e2f0f530, proc_io_chain=0x00007ffeeeb43930) at exec.cpp:573:17
   570 	            // We successfully made the attributes and actions; actually call
   571 	            // posix_spawn.
   572 	            int spawn_ret =
-> 573 	                posix_spawn(&pid, actual_cmd, &actions, &attr, const_cast<char *const *>(argv),
   574 	                            const_cast<char *const *>(envv));
   575
   576 	            // This usleep can be used to test for various race conditions

(const char *) actual_cmd = 0x00007ffeeeb435d1 "/bin/kill"
```

**Comment by zyv** [comment]
I'm updating this issue as it seems to be still reproducible with the latest mc (well, no wonder, not much has changed in our subshell interaction since then) and has become more prevalent with fish 4 as reported by @asl.

I think the suggestion in https://github.com/fish-shell/fish-shell/issues/6767 to use OSC-7 is a good one, but I'm afraid it won't solve the race problem, just probably lead to working directory desyncs instead of hangs, unless the races are found and fixed... or the subshell interaction is switched to an event loop, which might be a smaller and more self-contained project than going full on the event loop everywhere?

@krobelus, it seems like you did something on the fish side of things almost 5 years ago, so if it's possible to get you interested here, I'd be happy to work with you.

I'm also curious what @egmontkob would say about using OSC-7 instead of our own monkey business with PS1 hooks/emulation...

/cc @olfway after migrating to GitHub :)

**Comment by olfway** [comment]
I'm still able to reproduce it with 4.8.32 version from homebrew on m1 powerbook with macos 15.7

Also tried 2c8ed8dfe commit from https://github.com/MidnightCommander/mc/pull/4784

I'm able to reproduce it with:
1. open mc with "~/Library/Application Scripts" directory (it has ~1300 subfolders) 
2. in other tab or terminal run a script:
```
sleep 5; for i in (seq 1 250);
   echo -ne "\033]0;$i\007";
   osascript -e 'tell application "System Events" to key code 36';
end
```

You also have to allow It in "System Settings" (Privacy & Security -> Accessibility -> Terminal)

It usually hangs 3-4 times if I run this script 5 times, after it hangs I see subshell is stopped and kill is zombie:
```
2140 /opt/homebrew/Cellar/midnight-commander/4.8.32/libexec/bin/mc
2141 /opt/homebrew/bin/fish (stopped)
4854 (kill) (zombie)
```

To "unfreeze" it, you can send CONT and then STOP signals to subshell pid or WINCH signal to mc itself
or you can just resize terminal window

**Comment by egmontkob** [comment]
@olfway I only have ssh access to a mac box, so I guess `osascript` is out of question for me.

Websearch tells me key code 36 synthesizes an Enter (Return, whatever) key press – is this correct? So all you do is pressing Enter super quickly, switching back and forth between a directory and its parent?

As far as I understand, you give back the focus to your main terminal during the first 5 seconds, so that the synthesized keypresses go there? Is there any point in updating the title of the terminal where this script is running (no mc there)?

What fish version do you have?

In order to track down the problem I'd first need to somehow reproduce it, either on Linux, or on macOS over ssh. I don't know how I'll be able to do this.

**Comment by olfway** [comment]
@egmontkob yeah, you are correct

This scripts waits 5 seconds and then starts to send "Enter" quickly to active macos window

> So all you do is pressing Enter super quickly, switching back and forth between a directory and its parent?
yes, that's right

It updates title to make it visible on which sequence it right now, it's not important for reproducing bug

Actually, previously I did "Enter, Enter, Down" and then repeat. So to automate it I made this script.
But then I found that just "Enter" is enough to reproduce, so I simplified this script

But now, as you writing it, I realised that it's enough just to press on "Enter" on keyboard and wait it to freeze

I'm using current fish version from homebrew (4.1.2)

**Comment by olfway** [comment]
@egmontkob I'm able to reproduce it on different (empty) user with empty fish config and history, so it seems it's not because of some fish configuration or size of history

But it seems that it's important to have a folder with many folders/files inside, I used "~/Library/Containers", it has ~450 subfolders

**Comment by olfway** [comment]
I'm also able to reproduce it with zsh 5.9 (macos version) with almost empty config and history

I used "/Library/Developer/CommandLineTools/SDKs/MacOSX15.5.sdk/usr/share/man/man3" folder for that (it has 11523 files) and it freezes in one minute if I just keep "enter/return" key pressed on that directory

After "freeze" zsh subshell is in stopped state, and changing terminal size unfreezes it

**Comment by krobelus** [comment]
@olfway I tried to reproduce with your script with fish 4.1.2 and mc 4.8.33 on macOS Ventura (no configuration) inside `LIbrary/Application Scripts` but mc is not hanging.
So it's sending OSC 0 and Enter to mc 250 times; it visibly switches back and forth in mc, but no hang.

**Comment by asl** [comment]
@olfway @egmontkob There are all kinds of race conditions in subshell handling code. So if subshell is *fast enough* (like fish with empty config or so), then it is much easier to trigger these.

See explanation in https://github.com/MidnightCommander/mc/issues/4597#issuecomment-2695347964

I'm routinely hitting this issue locally. Resizing terminal sends signal, so it unfreezes `mc` (similar to the manual sending via `kill`).

**Comment by egmontkob** [comment]
Most race conditions, if we have a clue what triggers it, can pretty reliably be reproduced by adding a bit of delay to one of the components.

If the race condition occurs when the shell is too fast, let's add a delay to mc. If it happens when mc forwards the Enter keypress, let's add it after that action. This will make the shell even faster relative to mc, i.e. make the problem more likely to surface.

(If, on the other hand, the race appeared when mc is too fast then we could patch the shell to slow it down further in order to enlarge the race window.)

I'll give it a try.

**Comment by asl** [comment]
> Most race conditions, if we have a clue what triggers it, can pretty reliably be reproduced by adding a bit of delay to one of the components.

Yes, exactly. The path here seems to be:
 - `do_subshell_chdir` calls `feed_subshell`
 - `feed_subshell` calls `synchronize` in https://github.com/MidnightCommander/mc/blob/master/src/subshell/common.c#L932
 - Which waits for `SIGCHLD` in https://github.com/MidnightCommander/mc/blob/master/src/subshell/common.c#L545

What happens is that `SIGCHLD` is sent before the loop in https://github.com/MidnightCommander/mc/blob/master/src/subshell/common.c#L545 is entered. And therefore in absence of any other signals it just loops forever.

So, I'd probably suggest to add some delay before  `synchronize` call in https://github.com/MidnightCommander/mc/blob/master/src/subshell/common.c#L932

**Comment by egmontkob** [comment]
I tried inserting a sleep at various places, including before that `synchronize()`, but can't reproduce the problem on Linux.

Can anyone reproduce it on Linux? Or does it seem to be a macOS(/*BSD)-only thing?

> What happens is that SIGCHLD is sent before the loop

But, as far as I understand, the SIGCHLD then would set `subshell_stopped` and thus the loop wouldn't enter its `sigsuspend()`. The preceding `sigprocmask()` blocking SIGCHLD guarantees atomicity: the signal cannot hit after checking the `while` loop's condition but before entering `sigsuspend()`. I can't see a problem here.

---

PSA: Please don't link to blobs on `master` (or any other branch), always link to a particular tag or commit. The links you posted work _now_, but will stop working in the (probably near) future.

**Comment by asl** [comment]
> But, as far as I understand, the SIGCHLD then would set `subshell_stopped` and thus the loop wouldn't enter its `sigsuspend()`. The preceding `sigprocmask()` blocking SIGCHLD guarantees atomicity: the signal cannot hit after checking the `while` loop's condition but before entering `sigsuspend()`. I can't see a problem here.

I do not recall off-hand all the details currently – I tried to debug ~6 months ago and this is what I believe I saw, the signal was received earlier than expected and therefore missed.

**Comment by ossilator** [comment]
egmont's understanding is correct. so if there is a problem with this code, it must be some subtle implementation problem (maybe even a platform bug).

**Comment by egmontkob** [comment]
I'd much rather suspect an mc bug rather than a platform bug. After all, mc's code is quite complex and probably has quite a few race conditions. Just not exactly at that loop. The very purpose of `sigsuspend()`'s (which looks like a predecessor to `pselect()`) ability to install a new mask is to do it atomically, and it's clearly documented on macOS too, I doubt the problem would be here. Also the variables in the loop condition are properly marked as volatile. Locally, those few lines look okay to me.

Of course we could be having yet another platform difference like the tcdrain story was, but it's still an mc bug if it doesn't handle one of the platforms correctly.

Platform bug: Always possible, but I'd be surprised to find that here.

But this is just guesswork. We need to get back to investigating.

**Comment by asl** [comment]
I'll try to debug when I will have some spare time. Overall I'm seeing the issue maybe couple of times per hour or so locally :)

**Comment by egmontkob** [comment]
Just noticed (should've read the ticket carefully) that https://github.com/MidnightCommander/mc/issues/4071#issuecomment-2690107604 says that within the `while` loop, the value of `subshell_stopped` is something that it shouldn't be there. If the value was the same when `while`'s condition is evaluated, we wouldn't enter the loop. So it must be a signal handler (or another thread, but we're single threaded) modifying it.

That comment is from 2020, the code of `synchronize()` hasn't changed since 2011.

What I'd do next, on a platform that reproduces the issues, is to add debug `write(2, "A", 1);`-like statements (all with different letters) just before and after that `while`, inside the loop before and after that `sigsuspend()`, and to the beginning of `sigchld_handler`. `mc 2>log.txt` would redirect these to a log file. The order of the letters could confirm if the signal indeed hits around evaluating that `while` condition, i.e. the signal is not properly blocked. Or if maybe we enter `sigsuspend` twice, which would mean that the signal properly unblocks the first but then the `while` condition is erroneously evaluated to execute the body again, maybe the compiler mishandles 'volatile'.

@asl Could you please do this test for me? Its output would be very helpful.

**Comment by asl** [comment]
> maybe the compiler mishandles 'volatile'.

Reading the code once again... The variables are defined as follows:
```c
static volatile int subshell_alive, subshell_stopped;
```

Compiler might be ok here, but can be relaxed memory model on AArch64 cause some issues here? Though this would require signal handler and the loop to run on different cores... I would also try to change these to proper atomic to see if it makes the difference.

**Comment by egmontkob** [comment]
> can [the] relaxed memory model on AArch64 cause some issues here?

I am not familiar with these topics. I'd assume `volatile` to forbid the reordering of statements, but that might not be the case.

I'm not sure how making these two values atomic would improve the situation.

What I'm looking at is all those barrier, `memory_order_acquire`/`memory_order_consume`, `lwsync`, etc. stuff, but all I understand is that they're mostly C++, and that the C++ folks are still trying to figure out how to solve the related problems.

As for plain C, `__atomic_signal_fence()` (maybe without the leading underscores, [doc here](https://pubs.opengroup.org/onlinepubs/9799919799/functions/atomic_signal_fence.html), I'm yet to understand the details) and `__asm__ __volatile__("":::"memory");` seem to be relevant.

[Disabling optimization](https://stackoverflow.com/questions/2219829/how-can-i-prevent-gcc-optimizing-some-statements-in-c) for that one function might work, is worth a try. But, according to one of the answers there, `volatile` should be enough. [Another thread](https://stackoverflow.com/questions/22106843/gccs-reordering-of-read-write-instructions) disagrees with it.

---

Note that not only when entering that `while` loop, but also the end of the function is subject to a race condition if the compiler decides to freely shuffle around seemingly independent statements. First we unset `subshell_stopped`, and then we let the child continue and unblock SIGCHLD. If unsetting the boolean is moved to the end then a second SIGCHLD in quick succession can interleave there and effectively fail to set `subshell_stopped`, thus go missing.

---

By the way, just for fun, would be cool to see if gcc and clang both suffer from this issue, or maybe only one of them.

**Comment by olfway** [comment]
I tried to create a minimal app which reproduces issue and it seems that on macos (m1, with default clang 17) it randomly freezes on `sigsuspend` if `waitpid` was called in signal handler and if waitpid is called outside of signal handler then it works correctly.

On linux (amd64, gcc) it works correctly in both cases

test app tries to do SIGSTOP / SIGCONT with subprocess for 10000 times

I compile it with:
```
gcc -O2 -o subshell-test.exe subshell-test.c
```

and run with waitpid outside of handler:
```
$ ./subshell-test.exe
...
main: step=9999
main: set subshell_stopped = 0
main: sending STOP
main: before while loop
main: inside while loop, before sigsuspend

subshell: sending stop


sigchld_handler: begin
sigchld_handler: end

main: inside while loop, run check_subshell_status
check_subshell_status: pid == subshell_pid
check_subshell_status: signal = STOPPED, set subshell_stopped = 1
main: after sigsuspend, subshell_stopped is 1
main: after while loop, resume subshell
main: end of step=9999
```

with waitpid inside signal handler it freezes on random step:
```
$ ./subshell-test.exe handler
...
main: step=213
main: set subshell_stopped = 0
main: sending STOP
main: before while loop
main: inside while loop, before sigsuspend

subshell: sending stop


sigchld_handler: begin
sigchld_handler: run check_subshell_status
check_subshell_status: pid == subshell_pid
check_subshell_status: signal = STOPPED, set subshell_stopped = 1
sigchld_handler: end

main: after sigsuspend, subshell_stopped is 1
main: after while loop, resume subshell
main: end of step=213


main: step=214
main: set subshell_stopped = 0
main: sending STOP
main: before while loop

subshell: sending stop

main: inside while loop, before sigsuspend

sigchld_handler: begin
sigchld_handler: run check_subshell_status
check_subshell_status: pid == subshell_pid
check_subshell_status: signal = STOPPED, set subshell_stopped = 1
sigchld_handler: end

```

Here step 213 was successful and on step 214 handler was executed but main is still blocked on `sigsuspend`

[subshell-test.c](https://github.com/user-attachments/files/22922390/subshell-test.c)

**Comment by asl** [comment]
@olfway Interesting. Looks same here on M4:
```log
# ./a.out
subshell: sending stop


sigchld_handler: begin
sigchld_handler: end

main: inside while loop, run check_subshell_status
check_subshell_status: pid == subshell_pid
check_subshell_status: signal = STOPPED, set subshell_stopped = 1
main: after sigsuspend, subshell_stopped is 1
main: after while loop, resume subshell
main: end of step=9999

...

#./a.out handler
...

sigchld_handler: begin
sigchld_handler: run check_subshell_status
check_subshell_status: pid == subshell_pid
check_subshell_status: signal = STOPPED, set subshell_stopped = 1
sigchld_handler: end

main: after sigsuspend, subshell_stopped is 1
main: after while loop, resume subshell
main: end of step=9999
```

**Comment by asl** [comment]
UPD: So, it needs to be slower. Freezes `handler` when compiled `-O0`.

**Comment by asl** [comment]
Given that there is behavior difference in `-O2` vs `-O0` code I tried to see where I can insert some delays to trigger the erratic behavior with `-O2` code. It seems `usleep(1000)` in child loop (so between `read` and `raise(SIGSTOP)` makes `-O2`-compiled version to freeze somewhere:
```log
main: step=3845
main: set subshell_stopped = 0
main: sending STOP
main: before while loop
main: inside while loop, before sigsuspend

subshell: sending stop

main: after sigsuspend, subshell_stopped is 0
main: inside while loop, before sigsuspend

sigchld_handler: begin
sigchld_handler: run check_subshell_status
check_subshell_status: pid == subshell_pid
check_subshell_status: signal = STOPPED, set subshell_stopped = 1
sigchld_handler: end
```

**Comment by olfway** [comment]
@asl thanks for checking!

As far as I understand, mc currently uses the same approach with `sigsuspend` and `waitpid` in signal handler.

This means that if this test code doesn't contain any errors, then it's needed to remove the waitpid calls from the signal handler in mc or use a different approach altogether to make it work on macos reliable.

And it not depends on fish, bash or zsh at all

**Comment by asl** [comment]
> [@asl](https://github.com/asl) thanks for checking!
> And it not depends on fish, bash or zsh at all

It depends on the particular timeouts in the child, so might be implementation dependent in fact. Though this does not outline underlying problem. For now the main issue we do not understand why this combo (`sigsuspend` and `waitpid`) do not work, this seems to be a pretty standard way of signal handling.

**Comment by zyv** [comment]
FWIW, `waitpid()` seems to be explicitly allowed:

https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_04_03

> The following table defines a set of functions that shall be async-signal-safe. Therefore, applications can call them, without restriction, from signal-catching functions.

**Comment by asl** [comment]
@olfway another point – it seems that delays due to `write` calls are also significant. I cannot reproduce even with -O0 with debug printing turned off.

**Comment by asl** [comment]
I'm attaching ktrace dump (slightly edited process names for clarity).

[trace.txt.gz](https://github.com/user-attachments/files/22933562/trace.txt.gz)

The tracing terminated when the process was hanging on `sigsuspend`.

Have not analyzed it fully yet. The sizes of the character strings (`write` arguments, note that they are one-off due to `sizeof(msg) -1`) could be seen from the following LLVM IR dump (formatted ones excluded):
```llvm
@.str = private unnamed_addr constant [42 x i8] c"check_subshell_status: unknown pid value\0A\00", align 1
@.str.1 = private unnamed_addr constant [44 x i8] c"check_subshell_status: pid == subshell_pid\0A\00", align 1
@.str.2 = private unnamed_addr constant [67 x i8] c"check_subshell_status: signal = STOPPED, set subshell_stopped = 1\0A\00", align 1
@.str.3 = private unnamed_addr constant [48 x i8] c"check_subshell_status: skipping other statuses\0A\00", align 1
@.str.4 = private unnamed_addr constant [25 x i8] c"\0Asigchld_handler: begin\0A\00", align 1
@.str.5 = private unnamed_addr constant [44 x i8] c"sigchld_handler: run check_subshell_status\0A\00", align 1
@.str.6 = private unnamed_addr constant [23 x i8] c"sigchld_handler: end\0A\0A\00", align 1
@.str.7 = private unnamed_addr constant [8 x i8] c"handler\00", align 1
@.str.8 = private unnamed_addr constant [26 x i8] c"\0Asubshell: sending stop\0A\0A\00", align 1
@.str.10 = private unnamed_addr constant [32 x i8] c"main: set subshell_stopped = 0\0A\00", align 1
@.str.11 = private unnamed_addr constant [20 x i8] c"main: sending STOP\0A\00", align 1
@.str.12 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@.str.13 = private unnamed_addr constant [25 x i8] c"main: before while loop\0A\00", align 1
@.str.14 = private unnamed_addr constant [44 x i8] c"main: inside while loop, before sigsuspend\0A\00", align 1
@.str.15 = private unnamed_addr constant [52 x i8] c"main: inside while loop, run check_subshell_status\0A\00", align 1
@.str.16 = private unnamed_addr constant [47 x i8] c"main: after sigsuspend, subshell_stopped is 0\0A\00", align 1
@.str.17 = private unnamed_addr constant [47 x i8] c"main: after sigsuspend, subshell_stopped is 1\0A\00", align 1
@.str.18 = private unnamed_addr constant [41 x i8] c"main: after while loop, resume subshell\0A\00", align 1
```

**Comment by asl** [comment]
So, here is the last few lines of the log (annotation is mine):
```log
2025-10-15 11:30:54.319496 PDT          0.8              MACH_MKRUNNABLE # Signal handler is starting
2025-10-15 11:30:54.319496 PDT          0.2(6.1)         BSC_write  # main: inside while loop, before sigsuspend
2025-10-15 11:30:54.319498 PDT          0.0              BSC_write  # sigchld_handler: begin
2025-10-15 11:30:54.319499 PDT          0.0              BSC_write   # sigchld_handler: run check_subshell_status
2025-10-15 11:30:54.319500 PDT          0.0              BSC_wait4 # waitpid call
2025-10-15 11:30:54.319500 PDT          0.0              BSC_write  # check_subshell_status: pid == subshell_pid
2025-10-15 11:30:54.319501 PDT          0.0              BSC_write  # check_subshell_status: signal = STOPPED, set subshell_stopped = 1
2025-10-15 11:30:54.319502 PDT          0.0              BSC_wait4  # waitpid call 
2025-10-15 11:30:54.319502 PDT          0.0              BSC_write  # sigchld_handler: end
2025-10-15 11:30:54.319503 PDT          0.0              BSC_sigreturn  # return from signal handler
2025-10-15 11:30:54.319503 PDT          0.0              BSC_sigsuspend   # inside while loop
2025-10-15 11:30:54.319503 PDT          0.0              MACH_WAIT                            ad9e72e26254e5e1 2                0                0                28e303b            9(AP-P) parent(10497)
2025-10-15 11:30:54.319503 PDT          0.1              MACH_BLOCK                           0                fffffe0008d4d910 0                0                28e303b            9(AP-P) parent(10497)
2025-10-15 11:30:54.319503 PDT          0.0              MACH_SCHED_THREAD_SELECT             0                0                0                0                28e303b            9(AP-P) parent(10497)
2025-10-15 11:30:54.319503 PDT          0.2(0.2)         MACH_SCHED_THREAD_SELECT             2d4              0                0                a                28e303b            9(AP-P) parent(10497)
2025-10-15 11:30:54.319504 PDT          0.1              MT_InsCyc_CPU_CSwitch                1719678278080    938871c822ea     0                0                28e303b            9(AP-P) parent(10497)
2025-10-15 11:30:54.319504 PDT          0.0              MACH_STKHANDOFF                      0                2d4              1f               0                28e303b            9(AP-P) parent(10497)
```

it seems that the signal handler exited before `sigsuspend` call. Looks like what happens:

1. Parent signals child: `write(fd[1], "\n", 1);`
2. Parent enters the loop: `while (subshell_stopped == 0) {`
3. Here signal is sent by child, received by parent and signal handler is being run, `subshell_stopped` becomes 1
4. `sigsuspend` is called and waits for the *next* signal.

Looks like the race to me when the signal is received between checking loop condition and  `sigsuspend` call, no?

**Comment by olfway** [comment]
@asl I thought SIGCHLD signal should be blocked in the beginning with:
```
    sigemptyset(&mask);
    sigaddset(&mask, SIGCHLD);
    sigprocmask(SIG_BLOCK, &mask, &oldmask);
```

and only enabled while waiting with:
```
sigsuspend(&oldmask);
```

so I think signal handler shouldn't be run before sigsuspend

**Comment by asl** [comment]
> [@asl](https://github.com/asl) I thought SIGCHLD signal should be blocked in the beginning with:

@olfway Well, looks like it was not blocked somehow as you can see from the trace...

**Comment by ossilator** [comment]
i don't see a bug in the test program.

there is a [known issue](https://stackoverflow.com/questions/62252854/why-does-an-exiting-child-process-cause-the-parents-sigsuspend-call-to-return-o) with sigsuspend() on macos, through it shouldn't affect this case. but it might be worth a shot to check the signal mask after the sigsuspend() call.

**Comment by egmontkob** [comment]
I'll try to check the test program soon. But it looks like we're really suspecting some platform bug or nastiness that we'd have to work around.

One way to go would be to try the no optimization / memory barrier ideas from [this comment](https://github.com/MidnightCommander/mc/issues/4071#issuecomment-3396094405).

Another way to go would be to completely ditch the SIGSTOP - SIGCONT method of synchronizing. Switch to a different method. Let's say two pipes, in one of them the subshells sends a single byte to mc, in the other mc responds with a single byte (preferably a newline as that's the easiest to read from a shell).

Actually, the upwards part is already there when the subshell sends its pwd. So probably only a downwards pipe needed with a single byte sent.

**Comment by egmontkob** [comment]
For the record: `tcsh`'s equivalent to `read` is the `$<` construct, as in e.g. `set VAR = $<`, or simply `: $<` to throw away the data.

**Comment by egmontkob** [comment]
If using a [named pipe](https://github.com/MidnightCommander/mc/issues/4790) for synchronization, we don't even need to send a byte. Just either side opens it for writing and the other side opens it for reading, that'll sync them together. :)

**Comment by olfway** [comment]
> there is a [known issue](https://stackoverflow.com/questions/62252854/why-does-an-exiting-child-process-cause-the-parents-sigsuspend-call-to-return-o) with sigsuspend() on macos, through it shouldn't affect this case. but it might be worth a shot to check the signal mask after the sigsuspend() call.

@ossilator , that page have a link to another known issue:

[FB18565075](https://openradar.appspot.com/FB18565075): sigsuspend may return early without any signals being delivered, leaving the sigprocmask set incorrectly

Link is not working for me, so here is [webarchive](https://web.archive.org/web/20250928195830/https://openradar.appspot.com/FB18565075)

I also saw this previously and checked signal mask after sigsuspend - it was correct, but I've checked it only once. But it turns out that it's usually correct, but sometimes it's wrong and SIGCHLD is not blocked anymore

So I added 
```
sigprocmask(SIG_BLOCK, &mask, NULL);
```
right after `sigsuspend` call to reapply signal mask and now my test program doesn't breaks even with 1000000 repeats (previously it always breaks on 10000 repeats)

**Comment by olfway** [comment]
There is an interesting story on apple radar page about this bug which was fixed 32 years ago in BSD 4.4 but still present in latest macos unfortunately

**Comment by olfway** [comment]
> > [@asl](https://github.com/asl) I thought SIGCHLD signal should be blocked in the beginning with:
> 
> [@olfway](https://github.com/olfway) Well, looks like it was not blocked somehow as you can see from the trace...

@asl that bug also explains why it wasn't blocked

**Comment by egmontkob** [comment]
@olfway @asl Thanks a lot for your invaluable investigation!

I still haven't checked your code (and chances are that I won't need to), but as far as I can see:

We're confident that there's a platform bug.

My idea with memory barrier-like synchronization almost certainly won't fix it (it might make it accidentally less likely to occur due to different timings, but it's not the way to go).

You say

> signal mask after sigsuspend [...] usually correct, but sometimes it's wrong and SIGCHLD is not blocked anymore

> So I added sigprocmask(SIG_BLOCK, &mask, NULL); right after sigsuspend call to reapply signal mask

but in mc we do already have that sigprocmask() very shortly after sigsuspend() to unblock SIGCHLD, and yet we still have troubles. So maybe doing it much sooner just shrinks the race window?

Do you guys see a way to 100% reliably work around this macOS bug? Just following the thread from a distance, I can't.

If there's no 100% reliable workaround then we're left with two possibilities I guess:
- keep it buggy and blame Apple, hoping that they'll fix it soon,
- switch from signals to some other (most likely pipe-based) synchronization.

**Comment by zyv** [comment]
> Do you guys see a way to 100% reliably work around this macOS bug? Just following the thread from a distance, I can't.

Apparently there is a patch for `zsh`, which is suffering from the same problem:

```diff
diff --git a/Src/signals.c b/Src/signals.c
index de42f302d97f..b9c29e2f9216 100644
--- a/Src/signals.c
+++ b/Src/signals.c
@@ -228,8 +228,27 @@ signal_suspend(UNUSED(int sig), int wait_cmd)
      (sigtrapped[SIGINT] & ~ZSIG_IGNORED)))
    sigaddset(&set, SIGINT);
 
+#if defined(__APPLE__)
+    /* Work around https://openradar.appspot.com/FB18565075.
+     *
+     * Sometimes, sigsuspend returns early, leaving the process' signal mask
+     * set to its temporary mask argument without restoring the signal mask
+     * that was in effect prior to it being called. Save the signal mask before
+     * calling sigsuspend, and then manually restore it after sigsuspend
+     * returns.
+     */
+    sigset_t old_set;
+    int sigprocmask_rv = sigprocmask(SIG_SETMASK, NULL, &old_set);
+#endif
+
     ret = sigsuspend(&set);
 
+#if defined(__APPLE__)
+    if (sigprocmask_rv == 0) {
+        sigprocmask(SIG_SETMASK, &old_set, NULL);
+    }
+#endif
+
     return ret;
 }
```

https://superuser.com/a/1909498

> * keep it buggy and blame Apple, hoping that they'll fix it soon,

The bug was reported half a year ago, and existed for more than 3 decades. I'm somewhat skeptical that we could expect some reasonable reaction soon, and then there is of course a huge fleet of old machines...

**Comment by olfway** [comment]
I'm not sure about 100% reliable, but with sigprocmask() right after sigsuspend() I'm not able to freeze mc for a 30 minutes and without sigprocmask() it freezes in 2-3 minutes. (I cd to a folder with 10k files and put a heavy charger on "enter" key)

Current synchronize function:
```
{
    ...
    sigprocmask (SIG_BLOCK, &sigchld_mask, &old_mask);
    while (subshell_alive && !subshell_stopped)
        sigsuspend (&old_mask);

    if (subshell_state != ACTIVE)
    {
        tcflush (subshell_pty_slave, TCIFLUSH);
    }
    subshell_stopped = FALSE;
    kill (subshell_pid, SIGCONT);
    sigprocmask (SIG_SETMASK, &old_mask, NULL);
}
```

Yes, it's not doing too much until it restores old mask (and unblocks SIGCHLD anyway) in the end of function although setting `subshell_stopped = FALSE` and sending a signal to child `kill (subshell_pid, SIGCONT);` looks suspiciously and could break mc logic if signal handler will run in parallel

my current changes is:
```
    while (subshell_alive && !subshell_stopped) {
        sigsuspend (&old_mask);
#if defined(__APPLE__)
		sigprocmask (SIG_BLOCK, &sigchld_mask, NULL);
#endif
    }
```
as we already have a right `sigchld_mask` to apply and there is no need to save it before `sigsuspend`

So, I think it would be much better to apply similar fix anyway and maybe switch to pipe-based synchronization later

Freezing mc is a very bad user experience, especially if you don't know how to unfreeze it

**Comment by egmontkob** [comment]
@zyv ... but mc is already doing what that zsh patch adds as new code.

So either the patch is incomplete, or there's another bug somewhere...

Honestly, let's just ditch SIGSTOP / SIGCONT and synchronize via pipes! At least let's give it a try. It would also help address #4791.

---

One more thing we could try before giving up on signals is to see whether `pselect()` suffers from the same problem as `sigsuspend()`. As far as I understand, `pselect()` is a superset of `sigsuspend()`; if you don't specify fds or a timeout then it's the same (maybe the retval is different but we ignore that). _Maybe_ `pselect()` isn't buggy on mac.

**Comment by egmontkob** [comment]
@olfway You do have a point about that `kill(SIGCONT)`. Doesn't it also generate a SIGCHLD in turn? Maybe that causes further problems.

Instead of mac-specific `#if`, what if you simply swap the order of `kill()` and `sigsuspend()`, does that help?

**Comment by olfway** [comment]
> simply swap the order of kill() and sigsuspend(),

@egmontkob I'm not sure I get it. I tried to reapply SIGCHLD block right before `kill` but it not helps. It only helps if I re-apply mask right after `sigsuspend`. Although I didn't tried to reapply it right after `while` loop.

> Maybe pselect() isn't buggy on mac

Thats a great idea, I tried to replace `sigsuspend (&old_mask)` with `pselect(0, NULL, NULL, NULL, NULL, &old_mask)` and it seems it works reliably in both my test app and in mc! I wasn't able to freeze it in 20 minutes with this change

**Comment by egmontkob** [comment]
> I tried to reapply SIGCHLD block right before kill but it not helps. It only helps if I re-apply mask right after sigsuspend.

This suggests to me that there's probably still a race window, just much shorter than before.

> `pselect` [...] seems it works reliably

I'd suggest we go with that, then! Maybe, just to be on the safe side, could you do another huge batch of stress-testing? :)

> I think it would be much better to apply similar fix anyway and maybe switch to pipe-based synchronization later

Yup, if we seem to have a one-liner patch that works then let's go for that. We can switch to pipe-based later, much less urgently.

Thanks a lot for this amazing work!

**Comment by olfway** [comment]
@egmontkob 
I did a quick google and found that on some old platform (I didn't found exactly names) pselect is non-atomic userspace wrapper, so while it's a good solution for a modern systems, it's maybe a bit worse than atomic sigsuspend call for old ones.

But while searching, I found a description of "self pipe trick":

> On systems that lack pselect(), reliable (and more portable) signal	trapping can be achieved using the self-pipe trick:
> (where a signal handler writes a	byte to a pipe whose other end is monitored by select() in the main program.)

It seems we can use it in mc too, writing a byte to pipe in handler if a child process stopped and reading a byte instead of sigsuspend in synchronize function

**Comment by olfway** [comment]
@asl as you mention you have this issue several times per hour, could you also test pselect solution too, please?

It could be easily done with homebrew:

edit mc formula with:
```
brew edit midnight-commander
```

add:
```
  def pour_bottle?
    false
  end

  patch :DATA
```
right before `def install`, and a patch in the very end of formula file:
```
__END__
--- a/src/subshell/common.c
+++ b/src/subshell/common.c
@@ -542,7 +542,7 @@ synchronize (void)
 
     // Wait until the subshell has stopped
     while (subshell_alive && !subshell_stopped)
-        sigsuspend (&old_mask);
+        pselect(0, NULL, NULL, NULL, NULL, &old_mask);
 
     if (subshell_state != ACTIVE)
     {

```

and then, rebuild mc:
```
brew reinstall --build-from-source midnight-commander
```

**Comment by zyv** [comment]
Fantastic! Thanks a lot to @olfway, @asl for investigating and coming up with a solution, and also @krobelus & @egmontkob for participation.

Regarding the `pselect()` issue, from what I gather, the kernel implementation first appeared in Linux 2.6.16 (released on 20 March 2006) and before that it was emulated in glibc, suffering from the very same race condition it was supposed to fix.

I don't think we should care about systems using older kernels than that, unless the self-pipe trick is just as short and easy of a solution.

I'd appreciate a PR, only please do all a comment to the code summarizing the issue, otherwise one is left wondering why the change and it's all too easy to break it in the future.

**Comment by egmontkob** [comment]
pselect is POSIX since 2001. I wouldn't worry much about those systems that still emulate it in a non-atomic way (if any).

If you're really worried, you could do (pseudo-code)

```
#if apple
  /* briefly explain the problem, point to this issue & openradar */
  pselect
#else
  sigsuspend
#endif
```

(I wouldn't go for the self-pipe trick while keeping the signals; if that's desired then I'd much rather go straight for synchronizing with pipes only, no signals at all.)

**Comment by olfway** [comment]
> (I wouldn't go for the self-pipe trick while keeping the signals; if that's desired then I'd much rather go straight for synchronizing with pipes only, no signals at all.)

@egmontkob I've looked on how this self-pipe trick could be implemented and agree with you, it's better to keep signals with pselect/sigsuspend for now and later switch on signals-free synchronizing with pipes

**Comment by asl** [comment]
> [@asl](https://github.com/asl) as you mention you have this issue several times per hour, could you also test pselect solution too, please?
> 
> It could be easily done with homebrew:
> 
> edit mc formula with:
> 
> ```
> brew edit midnight-commander
> ```
> 
> add:
> 
> ```
>   def pour_bottle?
>     false
>   end
> 
>   patch :DATA
> ```
> 
> right before `def install`, and a patch in the very end of formula file:
> 
> ```
> __END__
> --- a/src/subshell/common.c
> +++ b/src/subshell/common.c
> @@ -542,7 +542,7 @@ synchronize (void)
>  
>      // Wait until the subshell has stopped
>      while (subshell_alive && !subshell_stopped)
> -        sigsuspend (&old_mask);
> +        pselect(0, NULL, NULL, NULL, NULL, &old_mask);
>  
>      if (subshell_state != ACTIVE)
>      {
> ```
> 
> and then, rebuild mc:
> 
> ```
> brew reinstall --build-from-source midnight-commander
> ```

Yes, I will try the `pselect` as soon as I will have a chance and report the results

**Comment by asl** [comment]
> Yes, I will try the `pselect` as soon as I will have a chance and report the results

@olfway @egmontkob I used it for few days and I doubt I've seen any hang so far!

**Comment by egmontkob** [comment]
Amazing news!

@olfway Could you please create a PR?

Please add a brief (perhaps one-line) comment to the code, summarizing this issue, so that a forthcoming contributor won't clean up the code to use the simpler `sigsuspend()` :)

As to whether use `pselect()` on all platforms, or make it specific to macOS only (are all BSDs fixed??) in the fear of `pselect()` not being atomic on some platforms, I have no preference; does anyone else have?

**Comment by egmontkob** [comment]
If we go for `pselect()` on all platforms then I'll file a new ticket to replace our self-pipe trick of `sigwinch_pipe` with `pselect()` 😄

**Comment by zyv** [comment]
> As to whether use `pselect()` on all platforms, or make it specific to macOS only (are all BSDs fixed??) in the fear of `pselect()` not being atomic on some platforms, I have no preference; does anyone else have?

I would usually err on the side of more ifdefs. However, in this case, the non-atomicity of `pselect` on Linux is over a 20-year-old story. Linux systems mostly don't live that long, and when they do, then they're not getting updates... So I'd rather use `pselect` here to keep the code cleaner with a comment referring to the unfixed state of `sigsuspend` on XNU as of now.

**Comment by ossilator** [comment]
yeah, we can go with pselect unconditionally.

@egmontkob i don't think that's a particularly good idea. we check that flag in quite some places. even if the check itself is centralized (i think it is), it's still a lot of extra syscalls.

**Comment by egmontkob** [comment]
> i don't think that's a particularly good idea

In order to try to keep things somewhat clean, I recommend that we discuss this in that yet-to-be-created issue. If we end up going with `pselect()` here for all platforms, that is.

**Comment by olfway** [comment]
I made a PR https://github.com/MidnightCommander/mc/pull/4805, switch to pselect on all platform

Please, review if you have any comments
